kale/embed/mpca.py

```diff
@@ -153,7 +153,8 @@ def _fit(self, x):
 
         # get the output tensor shape based on the cumulative distribution of eigenvalues for each mode
         for i in range(1, n_dims):
-            eig_values, eig_vectors = np.linalg.eig(phi[i])
+            # eig_values, eig_vectors = np.linalg.eig(phi[i])
+            eig_vectors, eig_values, vt = np.linalg.svd(phi[i])
```

Comment:
- SVD returns singular values rather than eigenvalues, so it is misleading here that you use variable `eig_values` for singular values (sqrt of eigenvalues).
- `phi[i]` should be `d x n` so I think it is correct to take the left singular vector but it will be good to verify it. I assume that you've did that, did you?

Reply:
- Will change. Is `u` `s` `vt` better?
- `phi[i]` in shape `I_i x I_i`, it is a sysmetric positive definite matrix, so theoretically `u` = `vt.T` for those vectors corresponding to non-zero singular values. I have checked this in debuging and we can consider this as a test case

Reply:
@sz144 This is **terribly wrong** and it is clear that you did not verify the math or code.  You should really do some unit testing after making such changes. Study the slides posted, you should do SVD on the data matrix X, but now you are doing SVD on X^TX. You should get a piece of paper to work it out and make sure you understand before coding.

This makes me aware that SVD may not be efficient because the _data matrix_ here is much larger than the co-variance matrix so it may take more time. You may study the time cost of both.

Reply:
In initial MATLAB implementation, we do eigen-decomposition for the **sum of covariance matrices**, but not the covariance of a certain matrix. So I do not know which matrix in `d x n` can be decomposed by SVD.

Reply:
That matrix is large. Considering mode-1: the matrix would be $I_1 \times (M \times I_2 \times \cdots \times I_N)$. And so we need to see how efficient SVD is performed on such a large matrix in numpy. We need to test it (for speed/space vs `eig`. If not, settling with `eig` may be better. `np.linalg.svd` does not allow to compute only `u` or only `vt`.

MPCA is doing X*X^T so getting the left singular vector is correct.

Reply:
The data matrix is effectively unfolding the whole (N+1)th-order data tensor along the nth mode.

Reply:
I see, thank you. I have tried `scipy.linalg.svd` where make `full_matrices=False`, and it works pretty fast. Memory error will be raised if `full_matrices=True`. We then need to compare/verify the results obtained by SVD and EIG.

Reply:
Got it. That's good to know. Yes, `full_matrices=False` is correct and it is not meaningful otherwise. That is a relief and if we implement this way, the code will be more compact too.